pkg/kubelet/dockertools/manager.go

```diff
@@ -1861,7 +1972,94 @@ func (dm *DockerManager) SyncPod(pod *api.Pod, _ api.PodStatus, podStatus *kubec
 		}
 	}
 
-	// Start everything
+	next, status, done := findActiveInitContainer(pod, podStatus)
+	if status != nil {
+		if status.ExitCode != 0 {
+			// container initialization has failed, flag the pod as failed
+			initContainerResult := kubecontainer.NewSyncResult(kubecontainer.InitContainer, status.Name)
+			initContainerResult.Fail(kubecontainer.ErrRunInitContainer, fmt.Sprintf("init container %q exited with %d", status.Name, status.ExitCode))
+			result.AddSyncResult(initContainerResult)
+			if pod.Spec.RestartPolicy == api.RestartPolicyNever {
+				glog.Errorf("Error running pod %q init container %q, restart=Never: %+v", format.Pod(pod), status.Name, status)
+				return
+			}
+			glog.Errorf("Error running pod %q init container %q, restarting: %+v", format.Pod(pod), status.Name, status)
+		}
+	}
+
+	if next != nil {
+		if len(containerChanges.ContainersToStart) == 0 {
+			glog.V(2).Infof("No containers to start, stopping at init container %+v in pod %v", next.Name, format.Pod(pod))
+			return
+		}
+
+		// If we need to start the next container, do so now then exit
+		container := next
+		startContainerResult := kubecontainer.NewSyncResult(kubecontainer.StartContainer, container.Name)
+		result.AddSyncResult(startContainerResult)
+
+		// containerChanges.StartInfraContainer causes the containers to be restarted for config reasons
+		// ignore backoff
+		if !containerChanges.StartInfraContainer {
```

Comment:
Why are we checking this here? the init containers are immutable so we should never ignore backoff right?